File: tests/vulnerable.py
import sqlite3
import os
import subprocess
import pickle
from flask import Flask, request

app = Flask(__name__)

# SQL Injection Vulnerability
def get_user(username):
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    # Vulnerable SQL query
    query = f"SELECT * FROM users WHERE username = '{username}'"
    cursor.execute(query)
    return cursor.fetchone()

# Command Injection Vulnerability
@app.route('/ping')
def ping_host():
    host = request.args.get('host', '')
    # Vulnerable command execution
    result = os.system(f"ping -c 1 {host}")
    return f"Ping result: {result}"

# Path Traversal Vulnerability
@app.route('/download')
def download_file():
    filename = request.args.get('file', '')
    # Vulnerable file access
    with open(filename, 'r') as f:
        return f.read()

# Insecure Deserialization
@app.route('/load_data')
def load_data():
    data = request.args.get('data', '')
    # Vulnerable deserialization
    return pickle.loads(data.encode())

# Hard-coded Credentials
def connect_to_database():
    username = "admin"
    password = "super_secret_password123"
    conn = sqlite3.connect('users.db', username=username, password=password)
    return conn

# Shell Injection
@app.route('/execute')
def execute_command():
    cmd = request.args.get('cmd', '')
    # Vulnerable shell command execution
    output = subprocess.check_output(cmd, shell=True)
    return output

# XSS Vulnerability
@app.route('/profile')
def show_profile():
    name = request.args.get('name', '')
    # Vulnerable XSS
    return f"<h1>Welcome, {name}!</h1>"

if __name__ == '__main__':
    app.run(debug=True)  # Debug mode enabled in production 

File: tests/Vulnerable.java
import java.io.*;
import java.sql.*;
import java.util.Base64;
import javax.servlet.http.*;
import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;

public class Vulnerable extends HttpServlet {
    // Hardcoded encryption key (vulnerability)
    private static final String ENCRYPTION_KEY = "MySuperSecretKey";
    
    // SQL Injection Vulnerability
    public ResultSet getUser(String username) {
        try {
            Connection conn = DriverManager.getConnection("jdbc:mysql://localhost/db", "root", "password");
            // Vulnerable SQL query
            String query = "SELECT * FROM users WHERE username = '" + username + "'";
            Statement stmt = conn.createStatement();
            return stmt.executeQuery(query);
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        }
    }
    
    // XSS Vulnerability
    protected void doGet(HttpServletRequest request, HttpServletResponse response) {
        try {
            String userInput = request.getParameter("input");
            PrintWriter out = response.getWriter();
            // Vulnerable XSS
            out.println("<html><body>Welcome, " + userInput + "!</body></html>");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    
    // Path Traversal Vulnerability
    public String readFile(String fileName) {
        try {
            // Vulnerable file access
            FileReader fr = new FileReader(fileName);
            BufferedReader br = new BufferedReader(fr);
            StringBuilder content = new StringBuilder();
            String line;
            while ((line = br.readLine()) != null) {
                content.append(line);
            }
            return content.toString();
        } catch (IOException e) {
            return "Error reading file";
        }
    }
    
    // Weak Encryption Implementation
    public String encryptData(String data) {
        try {
            // Using weak encryption
            Cipher cipher = Cipher.getInstance("DES/ECB/PKCS5Padding");
            SecretKeySpec key = new SecretKeySpec(ENCRYPTION_KEY.getBytes(), "DES");
            cipher.init(Cipher.ENCRYPT_MODE, key);
            return Base64.getEncoder().encodeToString(cipher.doFinal(data.getBytes()));
        } catch (Exception e) {
            return null;
        }
    }
    
    // Command Injection Vulnerability
    public String executeCommand(String command) {
        try {
            // Vulnerable command execution
            Process process = Runtime.getRuntime().exec(command);
            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            StringBuilder output = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }
            return output.toString();
        } catch (IOException e) {
            return "Error executing command";
        }
    }
    
    // Null Pointer Vulnerability
    public void processUser(String userId) {
        User user = null;
        // Vulnerable null pointer access
        if (userId.equals("admin")) {
            System.out.println("Admin user: " + user.getName());
        }
    }
    
    // Inner class with missing private data
    public class User {
        public String name;  // Should be private
        public String password;  // Should be private
        
        public String getName() {
            return name;
        }
    }
} 